pages/index.tsx

```diff
@@ -0,0 +1,169 @@
+import Head from "next/head";
+import styles from "../styles/Home.module.css";
+import React, { useEffect, useRef, useState } from "react";
+import fetch from "node-fetch";
+import * as jwt from "jsonwebtoken";
+import {
+  CLIENT_ID,
+  HOSTED_UI_LOGIN_URL,
+  HOSTED_UI_TOKEN_URL,
+  REDIRECT_URI,
+} from "./api/constants";
+import extractUrlParameter from "./api/extractUrlParameter";
+import generateSaml from "./api/generateSaml";
+import {
+  DecodedIdToken,
+  IdDetails,
+  SsoDetails,
+  TokenResponse,
+} from "./api/types";
+import CallCenterPicker from "../components/CallCenterPicker";
+import SelfSubmittingSsoForm from "../components/SelfSubmittingSsoForm";
+
+export default function Home() {
+  const [idDetails, setIdDetails] = useState<IdDetails | undefined>(undefined);
+  const [ssoDetails, setSsoDetails] = useState<SsoDetails | undefined>(
+    undefined,
+  );
+  const [pickerSubmitWasPressed, setPickerSubmitWasPressed] = useState(false);
+  const [singleSubmitTriggered, setSingleSubmitTriggered] = useState(false);
+  const [error, setError] = useState<string | undefined>(undefined);
+  // Control double effect-invocation when `fetch` should only happen once
+  const tokenFetchAlreadyRan = useRef<boolean>(false);
+
+  useEffect(() => {
+    // ensure `document` is going to be defined
+    if (typeof window !== "undefined") {
+      const authorizationCode = extractUrlParameter("code");
+
+      // No authorizationCode? Go directly to the Hosted UI login page
+      if (authorizationCode == null || authorizationCode === "") {
+        window.location.replace(HOSTED_UI_LOGIN_URL);
+      } else {
+        if (!tokenFetchAlreadyRan.current) {
+          tokenFetchAlreadyRan.current = true;
+          fetch(HOSTED_UI_TOKEN_URL, {
+            method: "POST",
+            headers: { "Content-Type": "application/x-www-form-urlencoded" },
+            body: new URLSearchParams({
+              grant_type: "authorization_code",
+              client_id: CLIENT_ID,
+              redirect_uri: REDIRECT_URI,
+              code: authorizationCode,
+            }),
+          })
+            .then((response) => {
+              if (response.ok) {
+                return response.json();
+              } else {
+                setError(
+                  "Status fetching Cognito session: " +
+                    `${response.status}. Have you tried logging in again?`,
+                );
+              }
```

Comment:
Same comment about simplifying if-else into an early exit if

Reply:
So this one is the one that made me do the other ones weird.
The `return` is the early exit, since the `setError` doesn't perform an exit really... not too sure what to do about that, but I guess I can drop the `else`.